contributors/design-proposals/storage/volume-reconstruct.md

```diff
@@ -0,0 +1,88 @@
+getting the existing pods information from pod manager. Reconciler can then
+recover actual state through reconcile loop. 
+
+The problem is that if pod is deleted during kubelet restarts, desired state
+no longer has this pod and volume information so as the actual state. But the
+actual mount points and direcotries are still exist on the disk. In order to
+clean up those left over mounts and directories, we propose a volume
+reconstruction process. The basic idea is to scan pod directories and recover
+volume information (such as volume spec) from the mount path and device mount
+path which will be updated into the actual state. With this reconstructure
+process, actual state can match the real world and reconciler will tear down
+volumes accordingly. 
+
+However, this reconstruction design has issues for some plugins such as ()
+because there is not enough information to reconstruct volume spec correctly.
+For those volume plugins, volume tear down will fail.
+
+## Possible solution
+
+In searching of the solutions, we first need to understand why we need to reconstruct volume spec. 
+Volume spec is an internal representation of a volume.  All API volume types translate to Spec.
+```
+type Spec struct {
+	Volume           *v1.Volume                // for inline volume use case
+	PersistentVolume *v1.PersistentVolume      // for pvc use case
+	ReadOnly         bool
+}
+```
+Both Volume and PersistentVolume has a name and volume source. Different volume plugins have different representation of their sources. 
+Volume spec is constructed when adding pod volume information into desired state.
+The main use of this spec is to get the volume source name to construct unique volume name (which typicall includes plugin name and volume source name). 
+The unique volume name is used as a key in actual/desired state. The volume source name is also used on construct global mount path (device mount path)
+for volume mounter. When volume is teared down, unmounter uses the same function to construct the mount path from the spec again.
+Different from global mount path, the bind mount path uses plugin name and spec name (not the volume source name). 
+
+So look more closely, you can see for only the purpose of volume tear down, if we have a record of those mount paths (bind and global) in actual state, there is no need to 
+construct those paths from the volume spec again. In case of kubelet restarts, those mount paths could be scanned from the mount lists left on the disk. (bind mount path has
+a path structure which can be easily idenetified and global mount path could be obtained from the mount reference of the bind mount path). This observation leads to our first
+possible solution.
+
+### Remove the dependency of volume spec. 
+Modify the data structure in actual state to keep a record of the mount paths so that volume tear down process (unmount) can use those records instead of reconstructing them
```

Comment:
volume tear down process is often not simple unmount.

- Ceph RBD needs to unmap ("node-initiated detach") the volume
- iSCSI needs to detach the volume (again, node-initiated detach)
- CSI can do virtually anything

All three need the full volume spec. Right now, we use json files placed on the filesystem with the volume spec. This brings its own issues, mainly with compatibility - new kubelet must be able to read files created by old kubelet(s) and it is not ideal solution either.

Reply:
The unmount device depends on Spec while unmount volume can be done without the spec. I am not sure, how this proposal is going to workaround the fact that, `UnmountDevice` requires full device spec.

Reply:
Maybe I should remove "(unmount)" here. I think I checked all plugins and from currently implementation, volume spec is only needed during volume set up (attach, waitforattach, mountdevice/mapdevice, setup). But for volume tear down (teardown, unmapdevice, TearDownDevice/unmountDevice, detach), only a couple of paths are needed (globalmountpath, mountpath/podDevicePath, devicePath)
For FC, Ceph RDB,  iSCSI: teardown includes unmapDevice (podDevicePath, globalMountPath), tearDownDevice(globalMountPath)
So I don't see they need full volume spec so far. Please let me know if I miss something.

Reply:
hmm, looking closer - I think I agree with @jingxu97 . Although volume spec is available in `GenerateUnmountDevice` func - we aren't using it. The `devicePath` is enough to unmount devices in most cases.

In case of iSCSI - we are saving disk configuration to a json file - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/iscsi/iscsi_util.go#L201 similar to CSI.

Reply:
@gnufied @jingxu97 

In case of iSCSI, we need many parameters(portal ip, iface name, iqn, initiatorname) to logout and cleanup iscsi interface.
All of them are stored in volumeSpec once, but it isn't provided during unmountDevice().
Therefore, iSCSI currently saves connection configuration into json file. If we get same parameters via volumeSpec at unmountDevice(), we don't need this logic.

Reply:
But as for the volume reconstruction path, iSCSI plugin can't create full spec using global mount path, therefore the plugin still need to store configurations in the json file same as CSI.

Reply:
@mtanino yes, they need those information during tear down. But we can get those information from plugin.Unmounter which only needs paratmers of InnerVolumeSpecName, and PodUID and we have those information. 
https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/iscsi/iscsi.go#L163
This unmounter only has podUID, volName, etc. But here https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/iscsi/iscsi_util.go#L432, during DetachDisk, I do see it might use Portals, Iqn,  Iface etc. Then what does it mean, those values will be just empty?